src/tss2-tcti/tcti-spi-lt2go.c

```diff
@@ -0,0 +1,278 @@
+        if (ret) {
+            if (retry++ > 5) {
+                LOG_ERROR ("libusb_bulk_transfer read failed with error: %s.", libusb_strerror(ret));
+                ret = TSS2_TCTI_RC_IO_ERROR;
+                goto out;
+            }
+            continue;
+        }
+        transfered += act_len;
+        length -= act_len;
+    }
+
+    if (data_in != NULL) {
+        memcpy(data_in, spi_dma_buffer, cnt);
+    }
+
+out:
+    memset (spi_dma_buffer, 0, SPI_MAX_TRANSFER);
+    return ret;
+}
+
+TSS2_RC
+platform_sleep_ms (void *user_data, int32_t milliseconds)
+{
+    (void) user_data;
+    struct timeval tv = {milliseconds/1000, (milliseconds%1000)*1000};
+    select (0, NULL, NULL, NULL, &tv);
+
+    return TSS2_RC_SUCCESS;
+}
+
+TSS2_RC
+platform_start_timeout (void *user_data, int32_t milliseconds)
+{
+     // Cast our user data
+    PLATFORM_USERDATA *platform_data = (PLATFORM_USERDATA *) user_data;
+
+    memset (&platform_data->timeout, 0, sizeof (struct timeval));
+
+    if (gettimeofday (&platform_data->timeout, NULL)) {
+        LOG_ERROR ("getimeofday failed with errno: %d.", errno);
+        return TSS2_TCTI_RC_GENERAL_FAILURE;
+    }
+
+    platform_data->timeout.tv_sec += (milliseconds/1000);
+    platform_data->timeout.tv_usec += (milliseconds%1000)*1000;
+    if (platform_data->timeout.tv_usec > 999999) {
+        platform_data->timeout.tv_sec++;
+        platform_data->timeout.tv_usec -= 1000000;
+    }
+
+    return TSS2_RC_SUCCESS;
+}
+
+TSS2_RC
+platform_timeout_expired (void* user_data, bool *is_timeout_expired)
+{
+     // Cast our user data
+    PLATFORM_USERDATA *platform_data = (PLATFORM_USERDATA *) user_data;
+
+    struct timeval now;
+    if (gettimeofday (&now, NULL)) {
+        LOG_ERROR ("getimeofday failed with errno: %d.", errno);
+        return TSS2_TCTI_RC_GENERAL_FAILURE;
+    }
+
+    if (now.tv_sec > platform_data->timeout.tv_sec) {
+        *is_timeout_expired = true;
+    }
+    if ((now.tv_sec == platform_data->timeout.tv_sec)
+        && (now.tv_usec > platform_data->timeout.tv_usec)) {
+        *is_timeout_expired = true;
+    }
+
+    *is_timeout_expired = false;
+    return TSS2_RC_SUCCESS;
+}
+
+void
+platform_finalize(void *user_data)
+{
+    // Cast our user data
+    PLATFORM_USERDATA *platform_data = (PLATFORM_USERDATA *) user_data;
+
+    if (platform_data->spi_dma_buffer)
+        libusb_dev_mem_free (platform_data->dev_handle, platform_data->spi_dma_buffer, SPI_MAX_TRANSFER);
+
+    if (platform_data->dev_handle)
+        libusb_close (platform_data->dev_handle); // close the device we opened
+
+    if (platform_data->ctx)
+        libusb_exit (platform_data->ctx); // needs to be called to end the
+
+    // Free user_data
+    free(platform_data);
+}
+
+TSS2_RC
+create_tcti_spi_lt2go_platform (TSS2_TCTI_SPI_HELPER_PLATFORM *platform)
+{
+    int ret = 0;
+    // Create required platform user data
+    PLATFORM_USERDATA *platform_data = malloc (sizeof (PLATFORM_USERDATA));
+
```

Comment:
no newline here, use calloc and drop memset below.

Reply:
fixed